### PR TITLE
Proximal Policy Optimization (contd) #1912

### PR DESCRIPTION
**Pull request by nishantkr18**
Hey everyone! After a long time, here's the continuation of the work from #1912. Hope we can finally have PPO in mlpack. (well, atleast for discrete action space environments, and when this works, we can think of adding the continuous version).

### CODE REVIEW

**Comment by nishantkr18** [comment]
@zoq I think the discrete version of PPO works now, atleast for most runs. Although the same implementation in pytorch works perfectly, on all runs. Could you have a look?
```python
# Building on top of the A2C code
from typing import Dict, List, Tuple
import matplotlib.pyplot as plt

import gym
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim


class Network(nn.Module):
    def __init__(self, in_dim: int, out_dim: int):
        """Initialization."""
        super(Network, self).__init__()

        self.layers = nn.Sequential(
            nn.Linear(in_dim, 128),
            nn.ReLU(),
            nn.Linear(128, out_dim)
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """Forward method implementation."""
        return self.layers(x)


class ppoAgent:
    """ppo Agent interacting with environment."""

    def __init__(
            self,
            env: gym.Env,
            learning_rate: int = 0.001,
            epsilon: float = 0.2,
            gamma: float = 0.99):

        obs_dim = env.observation_space.shape[0]
        action_dim = env.action_space.n

        self.env = env
        self.gamma = gamma
        self.epsilon = epsilon

        # Initialize the network
        self.actor = Network(obs_dim, action_dim)
        self.old_actor = Network(obs_dim, action_dim)
        self.old_actor.load_state_dict(self.actor.state_dict())
        self.critic = Network(obs_dim, 1)

        # Initialize optimizers
        self.optimizer_actor = optim.Adam(
            self.actor.parameters(), lr=learning_rate)
        self.optimizer_critic = optim.Adam(
            self.critic.parameters(), lr=learning_rate)

    def select_action(self, state: np.ndarray) -> np.ndarray:
        """Select an action from the input state."""
        logits = self.actor(torch.FloatTensor(state))
        prob = F.softmax(logits, dim=0).detach().cpu().numpy()
        return np.random.choice(len(prob), p=prob)

    def step(self, action: np.ndarray) -> Tuple[np.ndarray, np.float64, bool]:
        """Take an action and return the response of the env."""
        next_state, reward, done, _ = self.env.step(action)

        return next_state, reward, done

    def _t(self, array: np.ndarray) -> torch.FloatTensor:
        """Helper function to convert numpy to tensor"""
        return torch.FloatTensor(array)

    def train(self, steps: int):
        """Train the agent."""

        episode = 0
        scores = []
        score = 0

        state = self.env.reset()

        for step in range(1, steps + 1):
            action = self.select_action(state)
            next_state, reward, done = self.step(action)

            value = self.critic(self._t(state))

            advantage = reward + (1-done)*self.gamma * \
                self.critic(self._t(next_state)) - value
            advantage = advantage.detach()

            # Updating Critic
            loss_critic = -advantage * value
            self.optimizer_critic.zero_grad()
            loss_critic.backward()
            self.optimizer_critic.step()

            # Updating Actor
            logits = self.actor(self._t(state))
            log_prob = F.log_softmax(logits, dim=0)[action]
            with torch.no_grad():
                old_logits = self.old_actor(self._t(state))
                old_log_prob = F.log_softmax(old_logits, dim=0)[action]
        
            ratio = (log_prob - old_log_prob).exp()
            clipped = torch.clamp(
                ratio, 1-self.epsilon, 1+self.epsilon) * advantage
            log_prob_actions = torch.min(ratio * advantage, clipped)
            loss_actor = -log_prob_actions.mean()

            self.optimizer_actor.zero_grad()
            loss_actor.backward()
            self.optimizer_actor.step()
            self.old_actor.load_state_dict(self.actor.state_dict())

            state = next_state
            score += reward

            # if episode ends
            if done:
                episode += 1
                state = self.env.reset()
                scores.append(score)
                print("Step:", step, "\t Episode:", episode, "\t reward:",
                      score, "\t Avg reward: {:.2f}".format(
                          sum(scores)/len(scores)),
                      "\t Avg 100 rewards: {:.2f}".format(sum(scores[-100:])/len(scores[-100:])))
                if episode % 100 == 0:
                    self._plot(scores)
                score = 0

        self.env.close()

    def _plot(self, scores: np.ndarray):
        plt.figure(figsize=[12, 9])
        plt.subplot(1, 1, 1)
        plt.title(f"scores per episode")
        plt.plot(scores)
        plt.grid()

        # plt.show()
        plt.savefig('plots/4.ppo-discrete.png')
        plt.close()


if __name__ == "__main__":
    # environment
    env = gym.make("CartPole-v0")

    agent = ppoAgent(env)

    agent.train(200000)

    ### Testing agent ###
    state = env.reset()
    done = False
    score = 0

    while not done:
        env.render()
        action = agent.select_action(state)
        next_state, reward, done = agent.step(action)

        state = next_state
        score += reward

    print("score: ", score)
    env.close()
```

File: src/mlpack/methods/reinforcement_learning/ppo_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+  // Get action logits for each action at current state.
+  arma::colvec actionLogit;
+  actorNetwork.Predict(state.Encode(), actionLogit);
+
+  arma::colvec prob;
+  ann::Softmax<> softmax;
+  softmax.Forward(actionLogit, prob);
+
+  // Calculating cumulative probablity.
+  for (size_t i = 1; i<prob.n_rows; i++)
```
Comment (shrit) [inline src/mlpack/methods/reinforcement_learning/ppo_impl.hpp:131]:
```suggestion
  for (size_t i = 1; I < prob.n_rows; i++)
```

Comment (shrit) [inline src/mlpack/methods/reinforcement_learning/ppo_impl.hpp:131]:
```suggestion
  for (size_t i = 1; i < prob.n_rows; i++)
```

File: src/mlpack/methods/reinforcement_learning/ppo_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    prob[i] += prob[i-1];
+  
+  // Sampling action from cumulative probablity.
+  double randValue = arma::randu<double>();
+  for (size_t i = 0; i<prob.n_rows; i++)
```
Comment (shrit) [inline src/mlpack/methods/reinforcement_learning/ppo_impl.hpp:136]:
```suggestion
  for (size_t i = 0; i < prob.n_rows; i++)
```

File: src/mlpack/methods/reinforcement_learning/ppo_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  actorNetwork.Forward(sampledStates, actionLogit);
+  softmax.Forward(actionLogit, actionProb);
+  arma::mat prob = actionProb.row(action.action);
+
+  arma::mat ratio = prob / oldProb;
+
+  arma::mat L1 = ratio % advantages;
+  arma::mat L2 = arma::clamp(ratio, 1 - config.Epsilon(),
+                            1 + config.Epsilon()) % advantages;
+  // arma::mat surroLoss = -arma::min(L1, L2);
```
Comment (shrit) [inline src/mlpack/methods/reinforcement_learning/ppo_impl.hpp:209]: Any reason to keep this comment? since the `dSurroLos` is calculated

**Review by shrit — COMMENTED** [review]
Looks good to me, only some minor comment,
The implementation looks good, I am not an expert in PPO details, it represents perfectly the theoretical part.
@zoq Let me know what do you think, if everything looks good from your side, this one is ready to get merged

**Comment by zoq** [comment]
> Looks good to me, only some minor comment,
> The implementation looks good, I am not an expert in PPO details, it represents perfectly the theoretical part.
> @zoq Let me know what do you think, if everything looks good from your side, this one is ready to get merged

I'd like to run some checks to see why this fails on some runs.

**Pull request closed without merging**